### PR TITLE
Added/fixed extract. fields and opt. search func

### PR DESCRIPTION
**Pull request by catalinduta613**

### CODE REVIEW

File: Sniffer/main.py [hunk 1]
```diff
@@ -66,50 +68,62 @@ def packet_handler(frame):
...
         except:
             dbm_signal = "N/A"
-        logging.debug("dbm_signal: " + str(dbm_signal))
-
+        log("dbm_signal: " + str(dbm_signal))
+        now = datetime.datetime.now()
+        epoch_end = now.strftime("%H:%M")
+        # log(epoch_end)
+        epoch_minutes_ago = now - datetime.timedelta(minutes=EPOCH_LENGTH_IN_MIN)
+        epoch_start = epoch_minutes_ago.strftime("%H:%M")
```
Comment (adrianRoman24) [inline Sniffer/main.py:110]: not needed

File: Sniffer/main.py [hunk 2]
```diff
@@ -66,50 +68,62 @@ def packet_handler(frame):
...
                 "vendor" : vendor,
-                "vendor_id": vendor_id,
                 "supported_rates": supported_rates,
                 "channel": channel,
                 "dbm_signal": dbm_signal,
                 "ssid": ssid,
+                "supported_rates": supported_rates,
+                "ext_rates": ext_rates,
+                "epoch_end": epoch_end,
+                "epoch_start": epoch_start
```
[the remark this answers is not in the document]
Comment (adrianRoman24) [reply Sniffer/main.py:126]: i updated it to be stored here https://github.com/adrianRoman24/Monitoring-System-Randomization/blob/main/Sniffer/main.py#L124, so we use storage efficiently

File: Sniffer/main.py [hunk 3]
```diff
@@ -66,50 +68,62 @@ def packet_handler(frame):
...
+        log("supported_rates: "+ str(supported_rates))
+        # Extract the channel
         try:
-            channel = ord(frame[Dot11Elt][4].info)
+            channel = frame[Dot11].channel
         except:
             channel = "N/A"
-        logging.debug("channel: " + str(channel))
-
+        log("channel: " + str(channel))
```
Comment (adrianRoman24) [inline Sniffer/main.py:99]: use logging + debug

File: Sniffer/main.py [hunk 4]
```diff
@@ -51,8 +51,10 @@ def init_postgres_table():
     connection.commit()
     logging.info('Postgres table created')
 
+mac_to_index = {item["mac"]: i for i, item in enumerate(epoch_list)}
```
Comment (adrianRoman24) [inline Sniffer/main.py:54]: this is not correct, right? it is updated only once, when the program is started. You should just keep a dictionary with the mac addresses present in current epoch, when the epoch ends, reset the dictionary.

File: Sniffer/main.py [hunk 5]
```diff
@@ -66,50 +68,62 @@ def packet_handler(frame):
...
             dbm_signal = frame.dBm_AntSignal
         except:
             dbm_signal = "N/A"
-        logging.debug("dbm_signal: " + str(dbm_signal))
-
+        log("dbm_signal: " + str(dbm_signal))
+        now = datetime.datetime.now()
+        epoch_end = now.strftime("%H:%M")
+        # log(epoch_end)
+        epoch_minutes_ago = now - datetime.timedelta(minutes=EPOCH_LENGTH_IN_MIN)
```
Comment (adrianRoman24) [inline Sniffer/main.py:109]: EPOCH_LENGTH_IN_MIN does not exist anymore, use config. Did you even test the code?

**Review by adrianRoman24 — CHANGES_REQUESTED** [review]

File: Sniffer/main.py [hunk 6]
```diff
@@ -66,57 +70,62 @@ def packet_handler(frame):
...
-        except:
-            vendor_id = "N/A"
-
-        logging.debug("vendor_id: " + str(vendor_id))
-
+         # Extract the extended supported rates
+        ext_rates_ie = frame.getlayer(Dot11Elt, ID=50)
+        if ext_rates_ie:
+            ext_rates_raw = ext_rates_ie.info
+            ext_rates = [int(rate * 0.5) for rate in ext_rates_raw]
```
Comment (adrianRoman24) [inline Sniffer/main.py:85]: this variable is defined only when ext_rates_ie is not null, not ok

File: Sniffer/main.py [hunk 7]
```diff
@@ -66,57 +70,62 @@ def packet_handler(frame):
...
                 "mac" : source,
                 "timestamps" : [datetime.datetime.now()],
                 "vendor" : vendor,
-                "vendor_id": vendor_id,
                 "supported_rates": supported_rates,
                 "channel": channel,
                 "dbm_signal": dbm_signal,
                 "ssid": ssid,
+                "supported_rates": supported_rates,
+                "ext_rates": ext_rates,
```
Comment (adrianRoman24) [inline Sniffer/main.py:120]: this line throws error when the variable is not defined

File: Sniffer/main.py [hunk 8]
```diff
@@ -66,57 +70,62 @@ def packet_handler(frame):
...
+            ext_rates_raw = ext_rates_ie.info
+            ext_rates = [int(rate * 0.5) for rate in ext_rates_raw]
+            logging.debug("Extended Supported Rates:" + str(ext_rates))
+        else:
+            logging.debug("Extended Supported Rates: N/A")
+         # Extract the supported rates
         try:
-            supported_rates = frame[Dot11Elt][2].info.reaplace("'", '"')
+            supported_rates_raw = frame[Dot11Elt].getfieldval("rates")
+            supported_rates = [int((rate) * 0.5) for rate in supported_rates_raw]
```
Comment (adrianRoman24) [inline Sniffer/main.py:92]: what do the rates represent, what unit measure is this?

File: Sniffer/main.py [hunk 9]
```diff
@@ -51,8 +51,12 @@ def init_postgres_table():
     connection.commit()
     logging.info('Postgres table created')
 
+
 def check_new_detection(mac):
-    return next((i for i, item in enumerate(epoch_list) if item["mac"] == mac), None)
+    if mac in epoch_mac_dict:
```
Comment (adrianRoman24) [inline Sniffer/main.py:56]: you can use .get(..., None) method here, it is more efficient

**Review by adrianRoman24 — CHANGES_REQUESTED** [review]

File: Sniffer/main.py (hunk 8 above)
Comment (catalinduta613) [reply Sniffer/main.py:92]: "When the supported rates field is parsed from the packet, Scapy automatically maps the raw values to their corresponding human-readable representation. For example, the raw value of 2 corresponds to 1.0 Mbps, 4 corresponds to 2.0 Mbps, and so on. This mapping is based on the 802.11 standard, which defines the mapping between raw values and corresponding data rates."
So the values stored are in Mbps

**Review by adrianRoman24 — APPROVED** [review]
LGTM!

**Pull request merged**